projects/ADuCM3029_demo_cn0428_cn0429/src/ADuCM3029_demo_cn0428_cn0429.cpp

```diff
@@ -0,0 +1,1307 @@
+	if (pBuff[0] == 0) sprintf((char*)TXbuff,"Temp Comp = disabled%s", _EOS);
+	if (pBuff[0] == 1) sprintf((char*)TXbuff,"Temp Comp = enabled%s", _EOS);
+	UART_TX((const char*)TXbuff);
+	delay_ms(5);*/
+
+	eSensorResult = pGasSensor->close();
+	delay_ms(5);
+
+	return eSensorResult;
+}
+
+/*!
+ * @brief      Reads averaged ppb value from gas sensor
+ *
+ * @details    Reads sensor data utilizing custom gas sensor I2C library and sends this data over UART.
+ */
+int32_t SensorReadoutPPB(uint8_t sensor_address)
+{
+	int32_t PPB_Gas_Reading = 0;
+	eSensorResult = pGasSensor->openWithAddr(sensor_address);
+	eSensorResult = pGasSensor->ReadDataPPB(&PPB_Gas_Reading);
+	eSensorResult = pGasSensor->close();
+	delay_ms(10);
+	return PPB_Gas_Reading;
+}
+
+int32_t ExtractCfgVal()
+{
+	uint8_t buffer[16] = {0};
+	uint8_t read_char;
+	int i = 0;
+	for (i = 0; i < 16; i++) {
+		read_char = GetChar();
+		if (read_char == '\0') break;
+		else buffer[i] = read_char;
+	}
+	DeleteChar(i + 1);
+	return strtol((const char*)buffer, NULL, 10);
+}
+
+uint8_t *int2binString(int32_t a, uint8_t *buff, uint8_t bufSize)
+{
+	buff += (bufSize - 1);
+
+	for (int i = 31; i >= 0; i--) {
+		*buff-- = (a & 1) + '0';
+
+		a >>= 1;
+	}
+
+	return buff;
+}
+
+double Ieee754ConvertToDouble(uint8_t s[32])
+{
+	double f;
+	int16_t sign, exponent;
+	uint32_t mantissa;
+	int16_t i;
+
+	sign = s[0] - '0';
+
+	exponent = 0;
+	for (i = 1; i <= 8; i++)
+		exponent = exponent * 2 + (s[i] - '0');
+
+	exponent -= 127;
+
+	if (exponent > -127) {
+		mantissa = 1;
+		exponent -= 23;
+	} else {
+		mantissa = 0;
+		exponent = -126;
+		exponent -= 23;
+	}
+
+	for (i = 9; i <= 31; i++)
+		mantissa = mantissa * 2 + (s[i] - '0');
+
+	f = mantissa;
+
+	while (exponent > 0)
+		f *= 2, exponent--;
+
+	while (exponent < 0)
+		f /= 2, exponent++;
+
+	if (sign)
+		f = -f;
+
+	return f;
+}
+
+/* ================================================== */
+/* 					 State Machine					  */
+/* ================================================== */
+void DataDisplayFSM(void)
+{
+	uint8_t 	readChar = 0;	// variable holding address of selected sensor
+	uint8_t 	cmdResponseBuff[256] = {0};	// buffer holding sensor response to commands
+	uint8_t		tempBuff[40] = {0};			// buffer for EIS results parsing
+	double		EISpartialresult[8] = {0};	// array of partial results (one line) of EIS
+	int32_t		EISpartialDFTresult[8] = {0};	// array of partial results (DFT impedance) of EIS
+	uint32_t	EISresults[192] = {0};		// array of EIS results
+	int16_t 	sigVal = 0;		// temporary variable for signed value
+	uint16_t 	unsigVal = 0;	// temporary variable for unsigned value
+	uint32_t	unsiglVal = 0;	// temporary variable for long unsigned value
+	char		tempChar = 0;   // temporary variable for char
+	bool 		exitloop = false;
+
+	uint8_t pulseDuration, pulseAmplitude;	// Pulse test parameters
+
+	uint8_t tempError;
+	uint8_t numbytes;
+	uint8_t cmdLen;
+	uint16_t timeout;
+
+	switch(FSM_State) {
+	case INIT:
+		FSM_State = COMMAND;
```

Comment:
it's a good idea to put the states of a state-machine-function into their own separate function ; especially if the number of lines on each `case` statement go above a certain threshold

Reply:
Understood. Also in light of Mircea's similar comment below about the multiple if statements inside the case statement. I'll try pulling these out into separate functions.